eva/executor/create_index_executor.py

```diff
@@ -0,0 +1,160 @@
+
+        # Get feature tables.
+        table_ref = self.node.table_ref
+        df_metadata = CatalogManager().get_dataset_metadata(
+            table_ref.table.database_name, table_ref.table.table_name
+        )
+
+        # Add features to index.
+        # TODO: batch size is hardcoded for now.
+        index = None
+        input_dim, feat_size = -1, 0
+        for batch in StorageEngine.read(df_metadata, 1):
+            feat = batch.column_as_numpy_array(self.node.col_list[0].name)[0]
+            if index is None:
+                input_dim = feat.shape[-1]
+                index = self._create_index(self.node.index_type, input_dim)
+            index.add(feat)
+            feat_size += 1
+
+        # Input dimension is inferred from the actual feature.
+        input_index_io = CatalogManager().index_io(
+            "input_feature",
+            ColumnType.NDARRAY,
+            NdArrayType.FLOAT32,
+            [2],
+            True,
+        )
+
+        # Output dimension can be hardcoded.
+        id_index_io = CatalogManager().index_io(
+            "logical_id",
+            ColumnType.NDARRAY,
+            NdArrayType.INT64,
+            [2],
+            False,
+        )
+        distance_index_io = CatalogManager().index_io(
+            "distance",
+            ColumnType.NDARRAY,
+            NdArrayType.FLOAT32,
+            [2],
+            False,
+        )
+        save_file_path = (
+            EVA_DEFAULT_DIR
+            / INDEX_DIR
+            / Path("{}_{}.index".format(self.node.index_type, self.node.name))
+        )
+
+        io_list = [input_index_io, id_index_io, distance_index_io]
+
+        # Build secondary index to map from index Logical ID to Row ID.
+        # Use save_file_path's hash value to retrieve the table.
+        col_list = [
+            ColumnDefinition(
+                "logical_id",
+                ColumnType.INTEGER,
+                None,
+                [],
+                ColConstraintInfo(unique=True),
+            ),
+            ColumnDefinition(
+                "row_id", ColumnType.INTEGER, None, [], ColConstraintInfo(unique=True)
+            ),
+        ]
+        col_metadata = [
+            CatalogManager().create_column_metadata(
+                col.name, col.type, col.array_type, col.dimension, col.cci
+            )
+            for col in col_list
+        ]
+        tb_metadata = CatalogManager().create_metadata(
+            "secondary_index_{}_{}".format(self.node.index_type, self.node.name),
+            str(
+                EVA_DEFAULT_DIR
+                / INDEX_DIR
+                / Path("{}_{}.secindex".format(self.node.index_type, self.node.name))
+            ),
+            col_metadata,
+            identifier_column="logical_id",
+            is_video=False,
+        )
+        StorageEngine.create(tb_metadata)
+
+        # Write exact mapping for now.
+        logical_id = [i for i in range(feat_size)]
+        secondary_index = Batch(
```

Comment:
Why is it identity mapping?

Reply:
Logical ID is created according to the insertion order. We are currently building index row by row, so it is identity mapping for now. It may change if there is deletion or index update.

Reply:
I fear that the rows read from SQLite are not guaranteed to be ordered. So shouldn't we use some column from the table?

Reply:
Fixed this.